### ISSUE TITLE
couple thoughts

### ISSUE THREAD

**Issue by aspina7** [epiverse-trace/linelist#18]
Hey Thibaut - just read through as popped up on timeline, and had a few thoughts pop up. 
We already spoke briefly about this a while back but I like the concept and think is going to be super important to have an implementation like this that allows for generalising script templates (e.g. r4epis). 
We also talked briefly about whether a whole new class entirely necessary and thought I would just highlight a few things that would probably make the dev side easier and make it more useful for a wider audience of people: 
- have you checked out [{labelled}](https://github.com/larmarange/labelled)? similar to tags (and epis are used to this type of system from e.g. stata). Combined with {matchmaker}, is pretty strong way of standardising data with dictionaries. A simple implementation in the pr for [r4epis mortality templates](https://github.com/R4EPI/sitrep/pull/275/files#diff-f655e6e52d88cc56b4c4b3e88478d179ae223c760559f428df1b23cba82da967R822) - currently wip.  (It also plays nicely with {gtsummary} so clean output tables much easier. 
- in {epidict} we wrote a [function ](https://github.com/R4EPI/epidict/blob/master/R/msf_dict_rename_helper.R)for checking whether variables from a dictionary are used in a script 
- in terms of validation have you looked at {[pointblank](https://rich-iannone.github.io/pointblank/)}? 
- in {epikit} we have a [function](https://github.com/R4EPI/epikit/blob/master/R/find_date_cause.R#L176) to check that start date is before end date. 
- also wanted to remphasise that as a community we should work with ocha to create standardised hxl tags for common epi variables, and then a package like dirk's [{rhxl}](https://github.com/dirkschumacher/rhxl) would be a very lightweight solution when combined with {labelled} 

anyway, long story short - I like the idea and think as a community we should emphasise the importance of data dictionaries and build around that with existing infrastructure, and maybe deemphasise the need for separate class and infrastructure cascade. 
Just my two cents though :)

**Comment by aspina7** [comment]
> * the advantage of a dedicated class is easy dispatch to other methods we use for data processing/analysis and better input checking; definitely keen to minimize overhead on the user though, and hopefully the choice of relying on `data.frame`/`tibble` should make it more accessible

yeap - guessing one of the drawbacks is if being able to pass a dedicated class to other methods outside of the set used within this universe, though presumably if a `data.frame`/`tibble` that wouldnt be an issue? 

Shall we tag a discussion on to the beginning/end of the call re community?

**Comment by Bisaloo** [comment]
Thanks for the comments, Alex! It seems that discussion in this issue has come to and end. Let me reiterate some points for clarity as I close this issue:

> have you checked out [{labelled}](https://github.com/larmarange/labelled)? similar to tags (and epis are used to this type of system from e.g. stata). Combined with {matchmaker}, is pretty strong way of standardising data with dictionaries. A simple implementation in the pr for r4epis

As mentioned, this is now tracked in a dedicated issue: #19 

> - in {epidict} we wrote a [function ](https://github.com/R4EPI/epidict/blob/master/R/msf_dict_rename_helper.R?rgh-link-date=2022-04-25T21%3A03%3A43Z)for checking whether variables from a dictionary are used in a script
> - in terms of validation have you looked at {[pointblank](https://rich-iannone.github.io/pointblank/)}?
> - in {epikit} we have a [function](https://github.com/R4EPI/epikit/blob/master/R/find_date_cause.R?rgh-link-date=2022-04-25T21%3A03%3A43Z#L176) to check that start date is before end date.

These points are more related to data cleaning / validation, which is outside the scope of linelist. The [cleanepi package](https://github.com/epiverse-trace/cleanepi) should be what we are looking at in terms of data cleaning and validation, including cleaning of linelist objects.

> anyway, long story short - I like the idea and think as a community we should emphasise the importance of data dictionaries and build around that with existing infrastructure, and maybe deemphasise the need for separate class and infrastructure cascade.

Yes, good point, and I agree in general. I do think linelist address a somewhat unfilled niche, and is precisely a good candidate to integrate with the existing infrastructure. In particular, **it does not aim to replace data cleaning tools**, but to enhance them, and potentially facilitate the user interaction with these tools. As mentioned, the tagging system can be integrated with packages adding support for hxl tags. Similarly, linelist objects can be fed to tools like epikit, and use the tagging system to streamline the pipeline. For example, we could make sure this kind of things works:

``` r
library(epikit)
library(linelist)
library(outbreaks)

x <- make_linelist(
  measles_hagelloch_1861,
  id = "case_ID",
  date_onset = "date_of_prodrome",
  date_death = "date_of_death",
  age = "age",
  gender = "gender"
)

x |> 
  # Using tags here means this works with any column names in x
  assert_positive_timespan(has_tag("date_onset"), has_tag("date_of_death"))
#> Error in `assert_positive_timespan()`:
#> ℹ In argument: `has_tag("date_onset")`.
#> Caused by error:
#> ! `has_tag()` must be used within a *selecting* function.
#> ℹ See <https://tidyselect.r-lib.org/reference/faq-selection-context.html> for
#>   details.
```

<sup>Created on 2024-04-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

You're definitely welcome to open follow up issues as more questions or suggestions come to mind.
